geode-core/src/main/java/org/apache/geode/cache/query/security/RegExMethodAuthorizer.java

```diff
@@ -0,0 +1,167 @@
+    Objects.requireNonNull(cache, NULL_CACHE_MESSAGE);
+    Objects.requireNonNull(allowedPatterns, NULL_REGULAR_EXPRESSIONS_MESSAGE);
+
+    this.allowedPatterns = Collections.unmodifiableSet(allowedPatterns);
+    this.compiledPatterns = Collections.unmodifiableSet(compilePatterns());
+    this.restrictedMethodAuthorizer = new RestrictedMethodAuthorizer(cache);
+  }
+
+  /**
+   * Creates a {@code RegExMethodAuthorizer} object and initializes it so it can be safely used
+   * in a multi-threaded environment.
+   * <p/>
+   *
+   * @param restrictedMethodAuthorizer the default {@code RestrictedMethodAuthorizer} to use.
+   * @param allowedPatterns the regular expressions that will be used to determine whether a method
+   *        is authorized or not.
+   */
+  public RegExMethodAuthorizer(RestrictedMethodAuthorizer restrictedMethodAuthorizer,
+      Set<String> allowedPatterns) {
+    Objects.requireNonNull(allowedPatterns, NULL_REGULAR_EXPRESSIONS_MESSAGE);
+    Objects.requireNonNull(restrictedMethodAuthorizer, NULL_AUTHORIZER_MESSAGE);
+
+    this.allowedPatterns = Collections.unmodifiableSet(allowedPatterns);
+    this.compiledPatterns = Collections.unmodifiableSet(compilePatterns());
+    this.restrictedMethodAuthorizer = restrictedMethodAuthorizer;
+  }
+
+  /**
+   * Executes the authorization logic to determine whether the {@code method} is allowed to be
+   * executed on the {@code target} object instance.
+   * If the {@code target} object is an instance of {@link Region}, this methods also ensures that
+   * the user has the {@code DATA:READ} permission granted for the target {@link Region}.
+   * <p/>
+   *
+   * @param method the {@link Method} that should be authorized.
+   * @param target the {@link Object} on which the {@link Method} will be executed.
+   * @return {@code true} if the {@code method} can be executed on on the {@code target} instance,
+   *         {@code false} otherwise.
+   *
+   * @see org.apache.geode.cache.query.security.MethodInvocationAuthorizer
+   */
+  @Override
+  public boolean authorize(Method method, Object target) {
+    // Return false for known dangerous methods.
+    if (restrictedMethodAuthorizer.isKnownDangerousMethod(method, target)) {
+      return false;
+    }
+
+    // If the target object belongs to Geode, make sure the method is considered safe.
+    if (target.getClass().getPackage().getName().startsWith(GEODE_BASE_PACKAGE)) {
+      return restrictedMethodAuthorizer.isAllowedGeodeMethod(method, target);
+    }
+
+    // Compare fully qualified method name against compiled expressions.
+    String fullyQualifiedMethodName = target.getClass().getName() + "." + method.getName();
+    boolean match = compiledPatterns.stream()
+        .anyMatch(pattern -> pattern.matcher(fullyQualifiedMethodName).matches());
+
+    // Return if there was a match, otherwise delegate to the default authorizer.
+    return match || restrictedMethodAuthorizer.authorize(method, target);
```

Comment:
Just a small concern, for clarity it might be better to have each criteria that might be met (dangerous method, geode method, matching regex and default authorizer) have a separate return statement, rather than combining the last two into an or statement. Something like:

```
    if(compiledPatterns.stream().anyMatch(pattern -> pattern.matcher(fullyQualifiedMethodName).matches())) {
        return true
    }
    
    return restrictedMethodAuthorizer.authorize(method, target);
```

Reply:
I'll change it, thanks!.